### PR TITLE
[Bugfix] Fix request cancellation without polling

### LINKED ISSUE #10087: [Bug]: Requests aren't aborted when client disconnects if user provided a middleware
In https://github.com/vllm-project/vllm/pull/7111/, @njhill did a great job to make sure requests are aborted if they are cancelled by the client. This works great and can really ease the load on the model engine in cases of timeouts / other client disconnects.

Unfortunately, this mechanism doesn't work if a `BaseHTTPMiddleware` is added to the server, which is the case when a use provides a function to add as middleware (via the `--middleware` cmd arg of `vllm serve`). In this case, some weird starlette behavior causes `request.is_disconnected()` to return `False`, even if the client disconnects. This is discussed in https://github.com/encode/starlette/discussions/2094.
One of the suggested solutions in that discussion (for example, in https://github.com/encode/starlette/discussions/2094#discussioncomment-9084737 and https://github.com/encode/starlette/discussions/2094#discussioncomment-7576379) is to change the mechanism used to check if the request was actually disconnected. Copied verbatim from that discussion, use this helper instead of `request.is_disconnected()`:

> ``` python
> async def my_is_disconnected(request: Request) -> bool:
>     assert hasattr(request, '_is_disconnected'), "private API in starlette changed"
>     assert isinstance(request._is_disconnected, bool), "private API in starlette changed"
> 
>     if request._is_disconnected:
>         return True
> 
>     message: Message = {}
> 
>     # this timeout may need to be tweaked.
>     # Ideally request.receive() is non-blocking, in which case the timeout doesn't matter.
>     # But the ASGI spec seems to imply it should be a blocking callable, as it doesn't specify
>     # what should be returned in case of no new messages. In this situation you can return an
>     # empty dict, but given that it's against the spec it seems inadvisable.
>     with anyio.move_on_after(0.01):
>         message = await request.receive()
> 
>     if message.get("type") == "http.disconnect":
>         request._is_disconnected = True
> 
>     return request._is_disconnected
> 
> ```

In vLLM, the `is_cancelled` callable is passed to `iterate_with_cancellation` or to `merge_async_iterators`. In all cases, `request.is_disconnected` is used as that callable. So it should be pretty easy to replace it with the helper function instead. The only issue I see is with the timeout in `anyio.move_on_after`.

So WDYT? Is it possible to tweak the `is_cancelled` callable so requests will be aborted upon client disconnect also when the user provides a middleware function?

### PR DESCRIPTION
**Pull request by joerunde**
See discussion and alternate solution in #11096 

I came to agree with @jakkdl in https://github.com/encode/starlette/discussions/2094 and the linked issues that polling for disconnects with `request.is_disconnected()` introduces more problems than it's worth. Instead, we can use the pattern that's already in [StreamingResponse](https://github.com/encode/starlette/blob/0.41.3/starlette/responses.py#L252-L259) and have a separate async task wait for a disconnect message, cancelling our work if one is received. The key here is that a `StreamingResponse` is able to safely consume all new messages because the request body has already been read. Our request handlers have the same guarantee, since `fastapi` first reads and parses the request and builds a pdydantic object for us before invoking our handler.

This PR implements a decorator for our fastapi handlers that will cancel them if a disconnect message is received while they are running. This is implemented with `asynco` directly instead of with `anyio`, because the rest of the code base assumes `asyncio`.

The advantages here are:
- No extra polling overhead and need for configuring timeouts and poll intervals
- Work can be canceled as soon as a disconnect happens, instead of at the next disconnect check
- This is completely transparent to our code, and we can remove the existing cancelation handling that's coupled inside the server logic

Disadvantages
- This doesn't work with `entrypoints/api_sever.py` because that handler reads the request body itself, so our one cancellation test in `tests/async_engine/test_api_server.py` fails. I'll need to write a new one, but it's 6pm on a Friday 🙃 

I manually verified that this works to cancel both streaming and non-streaming requests, let me know what y'all think of doing this instead

FIX #10087

### CODE REVIEW

**Comment by joerunde** [comment]
Turns out the old async llm engine needed to check for cancellation errors for this new method to work, so I shuffled around the base `api_server.py` to hook up the new decorator so our test with it + the async llm engine works again.

Working now on a test for the openai api server, that doesn't necessarily need to block merging if we're on a tight schedule but should be in soon 🤞 

@mgoin @simon-mo Can I get a `ready` on this to get the full CI suite running?

**Comment by joerunde** [comment]
Okay, the test for request cancellation with the openai server is in as well. It overloads the server with a ton of requests and then cancels them, ensuring the server can still respond after. I would have rather been able to do something like check the `/metrics` output for aborted requests, but those aren't currently tracked by our metrics.

**Review by mgoin — APPROVED** [review]

File: tests/entrypoints/openai/test_basic.py
Comment (mgoin) [inline tests/entrypoints/openai/test_basic.py:144]: nit: Do we want to check that some number of requests (all?) have this error? While not likely, if all requests were successful we would still pass this section

**Review by mgoin — APPROVED** [review]
Even though it does have the potential to be a footgun (as you documented), I like the usage of `@with_cancellation` so LGTM pending green!

File: tests/entrypoints/openai/test_basic.py
Comment (joerunde) [reply tests/entrypoints/openai/test_basic.py:144]: This should already check that all of the requests fail, right? `pytest.raises` will throw an assertion error if the body of the context manager does not raise

Comment (mgoin) [reply tests/entrypoints/openai/test_basic.py:144]: Ah okay I misunderstood pytest, thanks!

**Comment by joerunde** [comment]
Ah shoot, the test was too much for the smaller cards we run on in CI and it failed even though the logs do show plenty of requests being aborted :(

I'll dial it back some- I don't want this to be flaky so I think fewer requests with more tokens each would be better at piling up load without burdening the server with handling so many aborts

**Pull request merged**
